### PR TITLE
KWCOCO Annotation Support

### LINKED ISSUE #285: COCO support
Import/export for https://cocodataset.org/#format-data

### PR DESCRIPTION
**Pull request by f4str**
fixes #285 

Link to sample data for testing: https://drive.google.com/file/d/1-WM7PMzTudP5BTF4VIQg6_MhseXMaVtb/view?usp=sharing

Added the ability to upload KWCOCO JSON files for annotations on web. The parser will automatically distinguish between this JSON and the DIVE JSON. COCO and KWCOCO have slight differences, but both are supported. 

There is full support for everything except attributes and track-attributes as they are user specific and to be decided later. Both image sequences and videos are supported with polygon and headl/tail keypoint annotations. Since DIVE only cares about head/tail keypoints, all other keypoints are ignored. 

Some exceptions include that RLE (run-length encoding) is not supported for polygons since DIVE only accepts the standard coordinate polygons. Furthermore, multiple polygons per annotation or polygons with holes are not supported due to lack of DIVE support. If the parser encounters any of these, it will throw an error and not parse the remainder of the file. 

For quick testing, a jupyter notebook exists in `notebooks/kwcoco-conversion.ipynb` to convert from KWCOCO JSON to DIVE JSON. Otherwise, simply uploading the images/video with the KWCOCO JSON normally will also work. 

Unit tests are also added in `server/tests/test_deserialize_kwcoco_json.py` which will be run with pytest and tox automatically. 

To create sample data for testing, the KWCOCO library can be used: [https://github.com/Kitware/kwcoco/blob/master/kwcoco/demo/toydata.py](https://github.com/Kitware/kwcoco/blob/master/kwcoco/demo/toydata.py) 

_Also related and slightly unrelated, it was determined that the direction at which the head/tail line points (from head to tail vs from tail to head) does not matter. The KWCOCO parser was created with this in mind, but the VIAME CSV parser was also modified to reflect this change in order to conform to consistency._

### CODE REVIEW

**Comment by subdavis** [comment]
This is looking awsome, I'll take a look really soon.  Thanks @f4str

File: server/dive_server/serializers/kwcoco.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+def annotation_info(
+    annotation: dict, meta: CocoMetadata
+) -> Tuple[int, str, int, List[int]]:
+    annotation_id = annotation['id']
+    image_id = annotation['image_id']
+    filename = meta.images[image_id]['file_name']
+
+    # may not exist, fall back to default values
+    trackId = annotation.get('track_id', annotation_id)
```
Comment (BryonLewis) [inline server/dive_server/serializers/kwcoco.py:32]: Looking at Kwcoco (https://github.com/Kitware/kwcoco#the-json-spec)  spec `track_id` could be:
```
    'track_id': <int | str | uuid>  # indicates association between annotations across images
```
Do we need to coerce this into an int somehow?

**Review by BryonLewis — CHANGES_REQUESTED** [review]
I need to go through it a bit more to look for more corner cases but this is really awesome.

One additional question:  Did you have some video based test-data that you used as well and if so could you provide that.  I was able to get the kwcoco toydata generated but I didn't dive too far into the commands for video generation.

File: server/dive_server/serializers/kwcoco.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+def is_coco_json(coco: Dict[str, Any]):
+    keys = ['categories', 'keypoint_categories', 'images', 'videos', 'annotations']
+    return any(key in coco for key in keys)
+
+
+def is_coco_string(coco: str):
+    data = json.loads(coco)
+    keys = ['categories', 'keypoint_categories', 'images', 'videos', 'annotations']
+    return any(key in data for key in keys)
```
[the remark this answers is not in the document]
Comment (f4str) [reply server/dive_server/serializers/kwcoco.py:21]: You're right. I wrote the `is_coco_string` function originally when I was reading the Girder File as a string (the same way the viame csv is parsed). However, I realized later that it's possible to get a file handler for the Girder File and wrote `is_coco_json`. Since `is_coco_string` is no longer being used, I think I should just get rid of the extraneous function.

File: server/dive_server/serializers/kwcoco.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+def _parse_annotation(
+    annotation: dict, meta: CocoMetadata
+) -> Tuple[dict, dict, dict, list]:
+    features: Dict[str, Any] = {}
+    attributes: Dict[str, Any] = {}
+    track_attributes: Dict[str, Any] = {}
+
+    category_id = annotation['category_id']
+    score = annotation.get('score', 1.0)  # may not exit, default to 1.0
```
[the remark this answers is not in the document]
Comment (f4str) [reply server/dive_server/serializers/kwcoco.py:51]: The `score` field does not always exist in COCO. 

Sometime is does like in this case: https://cocodataset.org/#format-results

Other times it does not like in this case: https://cocodataset.org/#format-data

[the remark this answers is not in the document]
Comment (BryonLewis) [reply server/dive_server/serializers/kwcoco.py:51]: I meant the comment there, it says `exit` not `exist`, sorry.

File: server/dive_server/serializers/kwcoco.py (hunk 1 above)
Comment (f4str) [reply server/dive_server/serializers/kwcoco.py:32]: In the case where the `track_id` is a string, it will most likely be in the form of a number in quote like "5" which is trivially easy to parse. If the type is UUID, things get bit more complicated. 

There is a python `uuid` module which can convert something like "cef0cbf3-6458-4f13-a418-ee4d7e7505dd" to a 128-bit integer but that number (although technically valid) would be massive so it would need to be reduced in some way. Another approach could be create a dictionary mappings of UUID to standard track numbers.

File: server/dive_server/serializers/kwcoco.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (f4str) [reply server/dive_server/serializers/kwcoco.py:51]: Ah I see. Yeah that's a typo, thanks for the catch.

File: server/dive_server/serializers/kwcoco.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                else:
+                    raise ValueError("Multiple polygons per annotation not supported")
+            else:
+                polygon = segmentation[0]  # get first polygon only
+
+            if isinstance(polygon, dict):  # dictionary kwcoco format
+                coords = polygon.get('exterior', [])
+                hole = polygon.get('interior', [])
+                if hole:
+                    raise ValueError("Polygon with hole not supported")
```
[the remark this answers is not in the document]
Comment (BryonLewis) [reply server/dive_server/serializers/kwcoco.py:107]: Rescinding my comment based on some further discussion.

**Review by subdavis — CHANGES_REQUESTED** [review]
Minor nits.  This looks really good.

File: server/dive_server/serializers/kwcoco.py (hunk 1 above)
Comment (subdavis) [reply server/dive_server/serializers/kwcoco.py]: I think I'd opt for something simple where UUIDs are converted to numbers by putting them all in a set, converting to to a list, and using the index of the UUID as the trackID.

**Review by BryonLewis — APPROVED** [review]
Sorry it took me so long to look at.  All the changes I asked for are good and thanks for providing the samples.

**Comment by subdavis** [comment]
If Bryon approves, no need to wait for me.

**Review by subdavis — APPROVED** [review]

**Comment by subdavis** [comment]
Merged because I wanted this to go into the release tomorrow morning.
